core/src/main/java/org/elasticsearch/index/translog/MultiSnapshot.java

```diff
@@ -65,4 +82,65 @@ public int totalOperations() {
     public void close() throws IOException {
         onClose.close();
     }
+
+    /**
+     * A wrapper of {@link FixedBitSet} but allows to check if all bits are set in O(1).
+     */
+    private static final class CountedBitSet {
+        private short onBits;
+        private final FixedBitSet bitset;
+
+        CountedBitSet(short numBits) {
+            assert numBits > 0;
+            this.onBits = 0;
+            this.bitset = new FixedBitSet(numBits);
+        }
+
+        boolean getAndSet(int index) {
+            assert index >= 0;
+            boolean wasOn = bitset.getAndSet(index);
+            if (wasOn == false) {
+                onBits++;
+            }
+            return wasOn;
+        }
+
+        boolean hasAllBitsOn() {
+            return onBits == bitset.length();
+        }
+    }
+
+    /**
+     * Sequence numbers from translog are likely to form contiguous ranges, thus using two tiers can reduce memory usage.
```

Comment:
I'm not familiar with the scale of the numbers here, but I take it that the memory usage is an important constraint? Without looking at how the sets all work in great detail, I guess this takes ~10x less memory - does that sound about right? Is that enough?

Reply:
I have adjusted the comment in https://github.com/elastic/elasticsearch/pull/27268/commits/59350092ea20dc6e7221409fe50633177e94f12b but not going to such detail.